scripts/release.sh

```diff
@@ -87,88 +105,226 @@ while true; do
 done
 
 # Check dependencies.
-dependencies gh sort
+dependencies gh jq sort
 
 if [[ -z $increment ]]; then
 	# Default to patch versions.
 	increment="patch"
 fi
 
+# Check if the working directory is clean.
+if ! git diff --quiet --exit-code; then
+	log "Working directory is not clean, it is highly recommended to stash changes."
+	while [[ ! ${stash:-} =~ ^[YyNn]$ ]]; do
+		read -p "Stash changes? (y/n) " -n 1 -r stash
+		log
+	done
+	if [[ ${stash} =~ ^[Yy]$ ]]; then
+		maybedryrun "${dry_run}" git stash push --message "scripts/release.sh: autostash"
+	fi
+	log
+fi
+
+# Check if the main is up-to-date with the remote.
+log "Checking remote ${remote} for repo..."
+remote_url=$(git remote get-url "${remote}")
+# Allow either SSH or HTTPS URLs.
+if ! [[ ${remote_url} =~ [@/]github.com ]] && ! [[ ${remote_url} =~ [:/]coder/coder(\.git)?$ ]]; then
+	error "This script is only intended to be run with github.com/coder/coder repository set as ${remote}."
+fi
+
 # Make sure the repository is up-to-date before generating release notes.
-log "Fetching $branch and tags from origin..."
-git fetch --quiet --tags origin "$branch"
+log "Fetching ${branch} and tags from ${remote}..."
+git fetch --quiet --tags "${remote}" "$branch"
 
 # Resolve to the latest ref on origin/main unless otherwise specified.
-ref=$(git rev-parse --short "${ref:-origin/$branch}")
+ref_name=${ref:-${remote}/${branch}}
+ref=$(git rev-parse --short "${ref_name}")
 
 # Make sure that we're running the latest release script.
-if [[ -n $(git diff --name-status origin/"$branch" -- ./scripts/release.sh) ]]; then
+script_diff=$(git diff --name-status "${remote}/${branch}" -- scripts/release.sh)
+if [[ ${script_check} = 1 ]] && [[ -n ${script_diff} ]]; then
 	error "Release script is out-of-date. Please check out the latest version and try again."
 fi
 
-# Check the current version tag from GitHub (by number) using the API to
-# ensure no local tags are considered.
-log "Checking GitHub for latest release..."
-versions_out="$(gh api -H "Accept: application/vnd.github+json" /repos/coder/coder/git/refs/tags -q '.[].ref | split("/") | .[2]' | grep '^v' | sort -r -V)"
-mapfile -t versions <<<"$versions_out"
-old_version=${versions[0]}
-log "Latest release: $old_version"
+# Make sure no other release contains this ref.
```

Comment:
It's a sanity check to prevent mistakes, do you see any harm in the check? Say you did `git rev-parse HEAD` on an old main, tried to create a release from that hash and it's already releases in version v2.10.0 but you're trying to create v2.11.0 (which will end up being older). We only want to go "forward".

Reply:
Yeah I'm thinking mainly about situations when this is expected. For instance, cut a emergency patch release using older revision.

Reply:
This is to safe-guard against making a mistake in such a case. It prevents you from creating v2.10.3 based on (v2.10.1+2, v2.10.2-2), only v2.10.2+X is possible.

Reply:
gotcha